### PR TITLE
Allow injection of Magento\Catalog\Model\View\Asset\ImageFactory

### PR DESCRIPTION
**Pull request by rolftimmermans**
It is not really possible to override the behaviour of Magento\Catalog\Model\View\Asset\ImageFactory.

Based on the suggestions given in https://github.com/magento/magento2/pull/9503#issuecomment-301195050 this implementation aims to fix that by allowing injection of a factory via the constructor.

### CODE REVIEW

**Comment by adragus-inviqa** [comment]
No need for an underscore for member names, yeah?

**Comment by rolftimmermans** [comment]
> No need for an underscore for member names, yeah?

I followed the conventions in the rest of the class...

File: app/code/Magento/Catalog/Model/Product/Image.php [hunk 1]
```diff
@@ -173,7 +173,7 @@ class Image extends \Magento\Framework\Model\AbstractModel
     /**
      * @var \Magento\Catalog\Model\View\Asset\ImageFactory
      */
-    private $viewAssetImageFactory;
+    protected $_viewAssetImageFactory;
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Image.php]: Could you revert this change? Variable names should not be prefixed with underscore according to PSR-2. Also protected variables should be replaced with private for all cases where it's possible

File: app/code/Magento/Catalog/Model/Product/Image.php [hunk 2]
```diff
@@ -226,6 +228,13 @@ public function __construct(
         $this->_assetRepo = $assetRepo;
         $this->_viewFileSystem = $viewFileSystem;
         $this->_scopeConfig = $scopeConfig;
+        if ($viewAssetImageFactory == null) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Image.php]: Would be great to replace this if block with ternary operators

**Comment by adragus-inviqa** [comment]
I know, but some of it is legacy/wrong. M2 uses [PSR-2](http://devdocs.magento.com/guides/v2.0/coding-standards/code-standard-php.html) for its code style, and that _suggests_ not to use underscores.

E.g. https://github.com/magento/magento2/pull/3401#discussion_r111334856

File: app/code/Magento/Catalog/Model/Product/Image.php (hunk 1 above)
Comment (okorshenko) [inline app/code/Magento/Catalog/Model/Product/Image.php]: please, revert this change. The property must be private

**Review by okorshenko — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/Product/Image.php [hunk 3]
```diff
@@ -226,6 +228,13 @@ public function __construct(
...
+            $this->_viewAssetImageFactory = ObjectManager::getInstance()->get(
+                \Magento\Catalog\Model\View\Asset\ImageFactory::class
+            );
+        } else {
+            $this->_viewAssetImageFactory = $viewAssetImageFactory;
```
Comment (orlangur) [inline app/code/Magento/Catalog/Model/Product/Image.php]: Please check this out: https://github.com/magento/magento2/blob/develop/app/code/Magento/Backend/Block/System/Account/Edit/Form.php#L60

- new dependency shall be added as last one
- there is no need in `if` statement - just make a one-liner
- always use imported class name like `OptionInterface::class`

**Review by orlangur — CHANGES_REQUESTED** [review]
Please read https://github.com/magento/magento2/pull/9503#issuecomment-300261918 more carefully.

So, even if you preferred inheritance for your customization and extended this class, you don't need to change property visibility to protected.

You may simply specify another class for this argument in `di.xml` for your new class.

**Comment by rolftimmermans** [comment]
I have made changes according to the requests by @okorshenko and @orlangur. Please let me know if this is sufficient to be considered for merging. Thanks!

**Review by ihor-sviziev — APPROVED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Review by okorshenko — APPROVED** [review]

**Review by orlangur — APPROVED** [review]

**Comment by magento-team** [comment]
@rolftimmermans thank you for your contribution to Magento 2 project
